synapse/util/task_scheduler.py

```diff
@@ -283,6 +283,20 @@ async def delete_task(self, id: str) -> None:
             raise Exception(f"Task {id} is currently ACTIVE and can't be deleted")
         await self._store.delete_scheduled_task(id)
 
+    def launch_task_by_id(self, id: str) -> None:
+        """Try launching the task with the given ID."""
+        # Don't bother trying to launch new tasks if we're already at capacity.
+        if len(self._running_tasks) >= TaskScheduler.MAX_CONCURRENT_RUNNING_TASKS:
+            return
+
+        run_as_background_process("launch_task_by_id", self._launch_task_by_id, id)
```

Comment:
The comment for this commit says `By making replication handler function non-async we avoid launching a
background task for each and every received command.`.
I don't really understand what we gain here and how it matches the commit comment. We are already calling `run_as_background_process` later on. This will do a first bg process that after some checks will do create another bg process and die. I don't think we are avoiding launching any bg task here but I may be missing something.

Reply:
We avoid launching the `run_as_background_process` if either a) the worker is not the background process (i.e. it won't run any tasks), and b) if we already have the maximum running tasks in parallel.

You're right that for the most part this doesn't really matter, but this does help in two important situations: a) it reduces CPU cost on other workers, and b) if we get lots and lots of new tasks we won't keep launching background tasks (which is useful when dealing with large spikes in traffic).

Reply:
> the worker is not the background process (i.e. it won't run any tasks)

This is already the case I think, `assert self._run_background_tasks` being the first line of `_launch_task` it never reaches `run_as_background_process` call.

> if we already have the maximum running tasks in parallel

Same here I think, the check is done in `_launch_task` before calling `run_as_background_process`.

Reply:
The thing is by declaring `on_NEW_ACTIVE_TASK` async we'll spawn a new background process automatically for every command that we get.

https://github.com/matrix-org/synapse/blob/7f801bc884f918d5a3622f520ac0c07c839063ec/synapse/replication/tcp/redis.py#L181-L194

Reply:
Ohhh ok.

Reply:
Yeah, its really subtle!